internal/querynode/impl.go

```diff
@@ -308,37 +308,27 @@ func (node *QueryNode) WatchDmChannels(ctx context.Context, in *querypb.WatchDmC
 		node: node,
 	}
 
-	err := node.scheduler.queue.Enqueue(task)
-	if err != nil {
-		status := &commonpb.Status{
-			ErrorCode: commonpb.ErrorCode_UnexpectedError,
-			Reason:    err.Error(),
-		}
-		log.Warn(err.Error())
-		return status, nil
-	}
-	log.Info("watchDmChannelsTask Enqueue done", zap.Int64("collectionID", in.CollectionID), zap.Int64("nodeID", Params.QueryNodeCfg.GetNodeID()), zap.Int64("replicaID", in.GetReplicaID()))
-	waitFunc := func() (*commonpb.Status, error) {
-		err = task.WaitToFinish()
+	future := node.channelPool.Submit(func() (interface{}, error) {
+		err := task.Execute(ctx)
 		if err != nil {
 			status := &commonpb.Status{
 				ErrorCode: commonpb.ErrorCode_UnexpectedError,
 				Reason:    err.Error(),
 			}
-			log.Warn(err.Error())
+			log.Warn("failed to subscribe dml channel ", zap.Error(err))
 			return status, nil
 		}
 
 		sc, _ := node.ShardClusterService.getShardCluster(in.Infos[0].GetChannelName())
 		sc.SetupFirstVersion()
-
-		log.Info("watchDmChannelsTask WaitToFinish done", zap.Int64("collectionID", in.CollectionID), zap.Int64("nodeID", Params.QueryNodeCfg.GetNodeID()))
+		log.Info("successfully watchDmChannelsTask", zap.Int64("collectionID", in.CollectionID),
+			zap.String("chan", in.Infos[0].GetChannelName()), zap.Int64("nodeID", Params.QueryNodeCfg.GetNodeID()))
 		return &commonpb.Status{
 			ErrorCode: commonpb.ErrorCode_Success,
 		}, nil
-	}
-
-	return waitFunc()
+	})
+	ret, err := future.Await()
+	return ret.(*commonpb.Status), err
```

Comment:
- Only gRPC error should be return here, another errors should be wrapped into the Status
- 'Await()` returning error may imply that ret is `nil`, then the type assertion will cause panic

Reply:
> * Only gRPC error should be return here, another errors should be wrapped into the Status
> * 'Await()`returning error may imply that ret is`nil`, then the type assertion will cause panic

If you take a look into it, the function itself will never return error.
Actually by our constraint, this function can only return a commonpb.Status.
Due to future definititon there must be error, but the error will always be nil.
change to return ret.(*commonpb.Status), nil should also do, work on it